casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -174,69 +249,130 @@ class RuntimeManagerImpl[F[_]: Concurrent] private[rholang] (
+
+  private def getData(
+      runtime: Runtime[F]
+  )(name: Par): F[Seq[Par]] =
+    runtime.space.getData(name).map(_.flatMap(_.a.pars))
+
+  private def getResult(
+      runtime: Runtime[F]
+  )(name: String = "__SCALA__"): F[Seq[Par]] =
+    runtime.space.getData(Par().withExprs(Seq(Expr(GString(name))))).map(_.flatMap(_.a.pars)) <*
+      computeEffect(runtime)(ConstructDeploy.sourceDeployNow(s"""for(_<-@"$name"){ Nil }"""))
 
   def getContinuation(
-      hash: ByteString,
-      channels: immutable.Seq[Par]
-  ): F[Seq[(Seq[BindPattern], Par)]] =
-    withResetRuntime(
-      hash,
-      runtime => {
-        runtime.space
-          .getWaitingContinuations(channels)
-          .map(
-            results =>
-              for {
-                result <- results.filter(_.continuation.taggedCont.isParBody)
-              } yield (result.patterns, result.continuation.taggedCont.parBody.get.body)
-          )
-      }
+      start: StateHash
+  )(channels: Seq[Par]): F[Seq[(Seq[BindPattern], Par)]] =
+    withResetRuntime(start)(
+      _.space
+        .getWaitingContinuations(channels)
+        .map(
+          _.filter(_.continuation.taggedCont.isParBody)
+            .map(result => (result.patterns, result.continuation.taggedCont.parBody.get.body))
+        )
     )
 
   private def newEval(
-      terms: Seq[DeployData],
-      runtime: Runtime[F],
-      initHash: StateHash
-  ): F[(StateHash, Seq[InternalProcessedDeploy])] = {
-
-    def doEval(
-        terms: Seq[DeployData],
-        hash: Blake2b256Hash,
-        acc: Seq[InternalProcessedDeploy]
-    ): F[(StateHash, Seq[InternalProcessedDeploy])] =
-      Concurrent[F].defer {
-        terms match {
-          case deploy +: rem =>
-            for {
-              _ <- runtime.space.reset(hash)
-              (codeHash, phloPrice, userId, timestamp) = ProtoUtil.getRholangDeployParams(
-                deploy
-              )
-              _                                      <- runtime.shortLeashParams.setParams(codeHash, phloPrice, userId, timestamp)
-              injResult                              <- doInj(deploy, runtime.reducer, runtime.errorLog)(runtime.cost)
-              EvaluateResult(evaluationCost, errors) = injResult
-              newCheckpoint                          <- runtime.space.createCheckpoint()
-              deployResult = InternalProcessedDeploy(
-                deploy,
-                Cost.toProto(evaluationCost),
-                newCheckpoint.log,
-                DeployStatus.fromErrors(errors)
-              )
-              cont <- if (errors.isEmpty)
-                       doEval(rem, newCheckpoint.root, acc :+ deployResult)
-                     else doEval(rem, hash, acc :+ deployResult)
-            } yield cont
-
-          case _ => (ByteString.copyFrom(hash.bytes.toArray), acc).pure[F]
+      runtime: Runtime[F]
+  )(deploys: Seq[DeployData], withPayment: Boolean): F[(StateHash, Seq[InternalProcessedDeploy])] =
```

Comment:
The cyclomatic complexity of this function makes it incomprehensible. Could you extract some clarifying functions?

Reply:
Porbably having a separate `newEval / computeState` for genesis would help. Then this one could always charge, and the 'genesis' one would never charge. Similarly, the genesis one could avoid a dozen checkpoints done in between the genesis deploys, most likely speeding up test setup and genesis block creation. It's "all or nothing" for genesis, after all.

Reply:
Yup, commenting out the mid-deploy checkpointing in newEval drops the RevVaultSpec startup time from 20s to 14s. So I guess having a separate 'genesisEval' is worth giving a shot for both performance and complexity reasons.

Reply:
@lukasz-golebiewski 

> The cyclomatic complexity of this function makes it incomprehensible. Could you extract some clarifying functions?

Can do.

@ArturGajowy 

> Porbably having a separate newEval / computeState for genesis would help.

I've considered this option, but it seemed sloppy. After all, that function would only be used once. It seems like those deploys should be evaluated in the constructor of RuntimeManager, and `emptyState` should be set to the result of those evaluations.

Reply:
At any rate, a separate function for initializing the manager would eliminate the need for the `withPayment` flag. That would greatly simplify this `newEval`.

Reply:
I'm not sure about the `emptyState`, but having two funcitons instead of one with a boolean flag would help greatly. Having one of them skip the in-the-middle checkpoints would further make things clearer, and faster at the same time.

Reply:
That introduces the issue that I presented last time we spoke. Casper doesn't discriminate between the genesis deploys and the deploys that follow. All of Casper assumes that the genesis deploys are validated like any other deploys, and so they're replayed like any other deploys. That means that we have to checkpoint them like any other deploys. Changing Casper to recognize the genesis deploys seems complex. I'd appreciate input from @KentShikama here.

Reply:
It looks like `computeBlockCheckpointFromDeploys` in `InterpreterUtil` [https://github.com/rchain/rchain/blob/05facc171003298ea95d5f640aa9082ac9708e93/casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala#L264](url) is a viable option for discriminating between genesis and non-genesis deploys. It calls `computeDeploysCheckpoint`, which calls `computeState`. This would be the place to call an alternative method that just ran the genesis deploys.

Reply:
That's a good one, in that it's actually only used in tests... I'll push a commit moving it to src/test.

Anyway, I don't think the intermediate checkpoints (hashes) are needed for replay. Are they?

Reply:
Nope, I won't. Please cherry-pick commit 31e83fb2 from my remote.

Reply:
We need the replay log for the genesis block because some genesis deploys are non-deterministic.

Reply:
@KentShikama thx for bringing that up. Can you point to concrete examples?

Reply:
Also, notice we didn't discuss disabling replay - we discussed disabling mid-deploy checkpointing for the genesis block. AFAIU the intermediate hashes / checkpoints are not needed in any way